checks/frozen_deps_test.go

```diff
@@ -91,3 +91,87 @@ func TestGithubWorkflowPinning(t *testing.T) {
 		})
 	}
 }
+
+func TestDockerfilePinning(t *testing.T) {
+	t.Parallel()
+	type args struct {
+		Filename string
+		Logf     func(s string, f ...interface{})
+	}
+
+	type returnValue struct {
+		Result bool
+		Error  error
+	}
+
+	l := log{}
+	tests := []struct {
+		name string
+		args args
+		want returnValue
+	}{
+		{
+			name: "Invalid dockerfile",
+			args: args{
+				Filename: "./testdata/Dockerfile-invalid",
+				Logf:     l.Logf,
+			},
+			want: returnValue{false, errors.New("file has no FROM keyword")},
+		},
+		{
+			name: "Pinned dockerfile",
+			args: args{
+				Filename: "./testdata/Dockerfile-pinned",
+				Logf:     l.Logf,
+			},
+			want: returnValue{true, nil},
+		},
+		{
+			name: "Pinned dockerfile as",
+			args: args{
+				Filename: "./testdata/Dockerfile-pinned-as",
+				Logf:     l.Logf,
+			},
+			want: returnValue{true, nil},
+		},
+		{
+			name: "Non-pinned dockerfile as",
+			args: args{
+				Filename: "./testdata/Dockerfile-not-pinned-as",
+				Logf:     l.Logf,
+			},
+			want: returnValue{false, nil},
+		},
+		{
+			name: "Non-pinned dockerfile",
+			args: args{
+				Filename: "./testdata/Dockerfile-not-pinned",
+				Logf:     l.Logf,
+			},
+			want: returnValue{false, nil},
+		},
+	}
+	for _, tt := range tests {
+		tt := tt // Re-initializing variable so it is not changed while executing the closure below
+		l.messages = []string{}
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			var content []byte
+			var err error
+
+			content, err = ioutil.ReadFile(tt.args.Filename)
+			if err != nil {
+				panic(fmt.Errorf("cannot read file: %w", err))
+			}
+
+			r, err := validateDockerfile(tt.args.Filename, content, tt.args.Logf)
+
+			if (err != nil && tt.want.Error == nil) ||
```

Comment:
You could simplify error checking logic by using `errors.Is`, i.e:

```
if !errors.Is(err, tt.want.Error) && r != tt.want.Result {
  t.Errorf()
}
```

Reply:
I initially did not do that because the error was created on the fly `return errors.New("file has no FROM keyword")` in `frozen_deps.go`. My intention was to clean this up once we handle errors properly.  
For now, I've created a new error `ErrInvalidDockerfile` at the top of `frozen_deps.go` and changed the unit test as you suggested. 
Long-term, are we going to have all errors in one file? Or at the top of individual files?

Reply:
> I initially did not do that because the error was created on the fly `return errors.New("file has no FROM keyword")` in `frozen_deps.go`. My intention was to clean this up once we handle errors properly.
> For now, I will create a new error `ErrInvalidDockerfile` at the top of `frozen_deps.go` and changed the unit test as you suggested.

You didn't need a new error here since you are not comparing error types using `errors.As` but instead the errors themselves using `errors.Is`. But if you already have defined a new error, thats great too.

> Long-term, are we going to have all errors in one file? Or at the top of individual files?

I think we should have them in a single file and not at top of individual files. If you are asking for this PR, its ok to keep it in the same file. We can move all errors out to a single file in a later PR.

Reply:
> > I initially did not do that because the error was created on the fly `return errors.New("file has no FROM keyword")` in `frozen_deps.go`. My intention was to clean this up once we handle errors properly.
> > For now, I will create a new error `ErrInvalidDockerfile` at the top of `frozen_deps.go` and changed the unit test as you suggested.
> 
> You didn't need a new error here since you are not comparing error types using `errors.As` but instead the errors themselves using `errors.Is`. 
Actually I tried that and it did not work when the error was created on the fly using `return errors.New("...")`

But if you already have defined a new error, thats great too.
> 
> > Long-term, are we going to have all errors in one file? Or at the top of individual files?
> 
> I think we should have them in a single file and not at top of individual files. If you are asking for this PR, its ok to keep it in the same file. We can move all errors out to a single file in a later PR.
sg